### PR TITLE
[DROOLS-2906] Implement scenario runner (backend)

### PR DESCRIPTION
**Pull request by danielezonca**
Implement new command to support pluggable external KieContainer and extended NewKieSessionCommand to support KieSession customization before creation

@Rikkola @jsoltes

PRs:
https://github.com/kiegroup/drools/pull/2036
https://github.com/kiegroup/droolsjbpm-knowledge/pull/323
https://github.com/kiegroup/drools-wb/pull/920

### CODE REVIEW

File: drools-core/src/main/java/org/drools/core/command/AbstractNewKieContainerCommand.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
+package org.drools.core.command;
```
Comment (Rikkola) [inline drools-core/src/main/java/org/drools/core/command/AbstractNewKieContainerCommand.java:17]: License

**Review by Rikkola — CHANGES_REQUESTED** [review]
Just checking. Is there unit test coverage for the changes?

File: drools-core/src/main/java/org/drools/core/command/AbstractNewKieContainerCommand.java (hunk 1 above)
Comment (danielezonca) [reply drools-core/src/main/java/org/drools/core/command/AbstractNewKieContainerCommand.java:17]: Done

**Comment by danielezonca** [comment]
@Rikkola 
Added two tests to cover new feature (new command and new customization)

**Review by Rikkola — APPROVED** [review]

File: drools-core/src/main/java/org/drools/core/command/AbstractNewKieContainerCommand.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    protected KieContainer getKieContainer(RegistryContext context, ReleaseId releaseId) {
+        KieContainer kieContainer;
+        if (releaseId != null) {
+            // use the new API to retrieve the session by ID
+            KieServices kieServices = KieServices.Factory.get();
+            kieContainer = kieServices.newKieContainer(releaseId);
+        } else {
+            kieContainer = context.lookup(KieContainer.class);
+            if (kieContainer == null) {
+                throw new RuntimeException("ReleaseId was not specfied, nor was an existing KieContainer assigned to the Registry");
```
Comment (jomarko) [inline drools-core/src/main/java/org/drools/core/command/AbstractNewKieContainerCommand.java]: Typo `specfied`.

Comment (danielezonca) [reply drools-core/src/main/java/org/drools/core/command/AbstractNewKieContainerCommand.java]: Done

Comment (jomarko) [reply drools-core/src/main/java/org/drools/core/command/AbstractNewKieContainerCommand.java]: Thanks

**Review by jsoltes — APPROVED** [review]

**Comment by danielezonca** [comment]
jenkins execute full downstream build

**Comment by danielezonca** [comment]
jenkins execute full downstream build

**Comment by danielezonca** [comment]
@manstis 
Ok after a rerun from 1 to 14 tests failed :)

**jBPM Random Failure Integration Tests**
- org.kie.server.integrationtests.jbpm.ProcessServiceIntegrationTest
- org.kie.server.integrationtests.jbpm.SLAComplianceIntegrationTest (x4)
- org.kie.server.integrationtests.jbpm.admin.ProcessInstanceAdminServiceIntegrationTest
- org.kie.server.integrationtests.jbpm.ProcessServiceIntegrationTest (x3)

**KIE Distribution Integration Tests: Scenario Asset is not present at all in KIE Distribution yet**
- org.kie.wb.test.rest.functional.ProjectIntegrationTest
- org.kie.wb.test.rest.functional.SpaceIntegrationTest
- org.kie.wb.test.rest.security.ProjectAccessIntegrationTest

**OptaPlanner Random Failure Integration Test**
- org.kie.server.integrationtests.optaplanner.jms.OptaPlannerJmsResponseHandlerIntegrationTest

**jBPM Failed Test: this is not a known random failure test so it is strange but it is a core jBPM test so I don't really think it is related**
- org.jbpm.test.container.test.ejbservices.tx.ETransactionTest

From my point of view none of them are related to these PRs. Wdyt?

**Comment by manstis** [comment]
Jenkins please retest this..

@danielezonca Just trying a build of this single repository...

Please let me know the outcome.. Sorry.

**Comment by manstis** [comment]
@mswiderski @yurloc Please see @danielezonca [comment](https://github.com/kiegroup/drools/pull/2036#issuecomment-419360247). IDK if you're aware of these failing tests.

@danielezonca Regarding the KIE WB Distribution test failures.. either fix the tests until the new asset is added or, I guess, add the test. We cannot have tests failing pending completion of some other work.

**Comment by danielezonca** [comment]
@manstis 
What I meant about failed KIE distribution integration tests is that all of them are part of the list of random failure tests and additionally at the moment none of new Scenario Test code is inside that project so it cannot be related.

**Comment by manstis** [comment]
@danielezonca Local build is green :-)

Just need @jomarko to approve and then these can all be merged!

**Review by jomarko — APPROVED** [review]

**Pull request merged**